test/lib/jdk/test/lib/format/ArrayDiff.java

```diff
@@ -0,0 +1,239 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.test.lib.format;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A difference between two arrays, which can be pretty formatted.
+ * For the calculated difference, user can request if the two arrays
+ * are equal (in terms of {@link Object#equals Object.equals()} for their
+ * elements). For the arrays that differ, a human-readable difference can
+ * be provided.
+ *
+ * <p>The difference is represented as a four-line text block, comprising of the
+ * first different element index, arrays printouts in the difference area,
+ * and a difference mark. For Primitive and Object elements in the source
+ * arrays their C-style escaped {@link String#valueOf String.valueOf()} are
+ * printed, element in String[] arrays are additionally surrounded with quotes.
+ * Additional formatting parameters, like maximum allowed width and number of
+ * elements printed before difference, can be specified.
+ *
+ * <p>Output examples:
+ *
+ * <p> two int arrays: </p>
+ * <pre>
+ * Arrays differ starting from [index: 4]:
+ *         ... 3, 4,   5, 6, 7]
+ *         ... 3, 4, 225, 6, 7]
+ *                  ^^^^
+ * </pre>
+ * <p> two String arrays: </p>
+ * <pre>
+ * Arrays differ starting from [index: 2]:
+ *         ["first", "second",     "third", "u\nprintable"]
+ *         ["first", "second", "incorrect", "u\nprintable"]
+ *                            ^^^^^^^^^^^^
+ * </pre>
+ * <p> two char arrays arrays: </p>
+ * <pre>
+ * Arrays differ starting from [index: 7]:
+ *         ... &#92;u0001, &#92;u0002, &#92;u0007, a, b, \n, ...
+ *         ... &#92;u0001, &#92;u0002,      }, a, b, \n, ...
+ *                            ^^^^^^^
+ * </pre>
+ */
+public class ArrayDiff<E> implements Diff {
+
+    private int failureIdx;
+    private final int maxWidth;
+    private final int contextBefore;
+
+    private final ArrayCodec<E> first;
+    private final ArrayCodec<E> second;
+
+    /**
+     * Default limits for the formatter
+     */
+    public static class Defaults {
+        final static int WIDTH = 80;
+        final static int CONTEXT_BEFORE = 2;
```

Comment:
either these constants should be `public`, or `Defaults` class should be `package-private`, otherwise, you get a public class w/ no public fields.

Reply:
Old cpp sins... fixed.